gflownet/envs/spacegroup.py

```diff
@@ -0,0 +1,335 @@
+        """
+        if state is None:
+            state = self.state
+        crystal_system_idx = state[self.cs_idx]
+        if crystal_system_idx != 0:
+            crystal_system = self.crystal_systems[crystal_system_idx][0]
+        else:
+            crystal_system = "None"
+        point_symmetry_idx = state[self.ps_idx]
+        if point_symmetry_idx != 0:
+            point_symmetry = self.point_symmetries[point_symmetry_idx][0]
+        else:
+            point_symmetry = "None"
+        space_group = state[self.sg_idx]
+        if space_group != 0:
+            point_group = self.space_groups[space_group][0]
+            crystal_class_idx = self.space_groups[space_group][1]
+            crystal_class = self.crystal_classes[crystal_class_idx][0]
+        else:
+            # TODO: Technically the point group and crystal class could be determined
+            # from crystal system + point symmetry
+            point_group = "TBD"
+            crystal_class_idx = 0
+            crystal_class = "TBD"
+        readable = (
+            f"{space_group} | {crystal_system} ({crystal_system_idx}) | "
+            + f"{crystal_class} ({crystal_class_idx}) | {point_group} | "
+            + f"{point_symmetry} ({point_symmetry_idx})"
+        )
+        return readable
+
+    def readable2state(self, readable):
+        """
+        Converts a human-readable representation of a state into the standard format.
+        See: state2readable
+        """
+        properties = readable.split(" | ")
+        crystal_system = int(properties[1].split(" ")[-1].strip("(").strip(")"))
+        point_symmetry = int(properties[4].split(" ")[-1].strip("(").strip(")"))
+        space_group = int(properties[0])
+        state = [crystal_system, point_symmetry, space_group]
+        return state
+
+    def reset(self, env_id=None):
+        """
+        Resets the environment.
+        """
+        self.state = self.source.copy()
+        self.n_actions = 0
+        self.done = False
+        self.id = env_id
+        return self
+
+    def get_parents(self, state=None, done=None, action=None):
+        """
+        Determines all parents and actions that lead to a state.
+
+        Args
+        ----
+        state : list
+
+        done : bool
+            Whether the trajectory is done. If None, done is taken from instance.
+
+        action : None
+            Ignored
+
+        Returns
+        -------
+        parents : list
+            List of parents in state format
+
+        actions : list
+            List of actions that lead to state for each parent in parents
+        """
+        if state is None:
+            state = self.state.copy()
+        if done is None:
+            done = self.done
+        if done:
+            return [state], [(self.eos, 0)]
+        else:
+            parents = []
+            actions = []
+            # Catch cases where space group has been selected
+            if state[self.sg_idx] != 0:
```

Comment:
So if I understand correctly, here we consider the cases where: 1) we went from an empty state to a defined crystal system / point symmetry in a single step via setting space group, and 2) where we are "unsetting" space group only (and crystal system / point symmetry remain as is). But in the second case, crystal system and point symmetry will always be non-zero (since they will be set to something in step based on the space group). And if so, aren't we missing the case in which in the first step we set either the crystal system or the point symmetry (but not both - the other remains zero), and in the second step we set space group?

Reply:
Let me start by restating your question:
> aren't we missing the case in which in the first step we set either the crystal system or the point symmetry (but not both - the other remains zero), and in the second step we set space group?

My answer is no, because I believe this is captured in the for loop in (current) [line 289](https://github.com/alexhernandezgarcia/gflownet/blob/spacegroup/gflownet/envs/spacegroup.py#L284):
```
for prop, idx in enumerate(state[: self.sg_idx]):
	if idx != 0:
		parent = state.copy()
		parent[prop] = 0
		parents.append(parent)
		action = (prop, idx)
		actions.append(action)
```
If the space group property is non-zero (`state[self.sg_idx] != 0`), then you are right that the other two properties (crystal system and point symmetry must be non-zero too because they will have been set by `step()`. But therefore, inside the above for loop, the condition `if idx != 0` will be true for both properties and both parents will be appended to the list of parents.

Please let me know if you think I am wrong.

Reply:
You are right, sorry, my bad - I think I might have missed `state[self.sg_idx] = 0` or something.